src/bootstrap/check.rs

```diff
@@ -115,7 +107,35 @@ pub fn cargo(build: &Build, stage: u32, host: &str) {
     // available.
     cargo.env("CFG_DISABLE_CROSS_TESTS", "1");
 
-    build.run(cargo.env("PATH", newpath));
+    build.run(cargo.env("PATH", &path_for_cargo(build, compiler)));
+}
+
+/// Runs `cargo test` for the rls.
+pub fn rls(build: &Build, stage: u32, host: &str) {
+    let compiler = &Compiler::new(stage, host);
+
+    let mut cargo = build.cargo(compiler, Mode::Tool, host, "test");
+    cargo.arg("--manifest-path").arg(build.src.join("src/tools/rls/Cargo.toml"));
+
+    // Don't build tests dynamically, just a pain to work with
+    cargo.env("RUSTC_NO_PREFER_DYNAMIC", "1");
+
+    build.add_rustc_lib_path(compiler, &mut cargo);
+
+    // Avoid using the rustc shim.
+    cargo.env_remove("RUSTC");
```

Comment:
Er wait hang on, I think I may have spoken too soon!

So Cargo's a pretty special flower here. I'm remembering now that we *do* use the shim rustc to compile Cargo itself. This correctly accounts for things like optimization flags, debuginfo flags, etc. When we run tests, however, Cargo *would* use the `RUSTC` env var configured but all of Cargo's tests actually [explicitly ignore this env var](https://github.com/rust-lang/cargo/blob/9fcdbb44fec18d10c086b6aba8143bc06a199761/tests/cargotest/lib.rs#L55) relying on `rustc` instead. That means that the `PATH` modification is purely done for tests and `RUSTC` is purely done for compiling Cargo itself. Weird, no?

So now we've got a question for the rls itself. I believe the rls *should* be compiling with `RUSTC` (the shim) actually to get the right flags, so I believe I was mistaken to request this `env_remove`. Down below though it still seems that the `PATH` modification should not be necessary. Does the test suite for the rls tamper with `RUSTC` at all? If it's invoking Cargo and/or respecting `RUSTC` then I'd actually expect the rls to basically ignore the `PATH` below, always relying on `RUSTC` itself (a different compiler).

You mentioned errors with the bitflags crate, but some of those may have actually been very recently fixed, so maybe a rebase, remove `env_remove`, remove `PATH` changes, and see if it works?